CARDS-1881: Periodical export task that serializes new forms as CSVs

sofi2002sofi opened the pull request:
[CARDS-1881](https://phenotips.atlassian.net/browse/CARDS-1881)

To test:
On http://localhost:8080/system/console/configMgr add a new item of "Survey csv export". By default, the task will be invoked each day at midnight - 12 am, and saved under the project root directory.

[CARDS-1881]: https://phenotips.atlassian.net/browse/CARDS-1881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportEndpoint.java:
```diff
@@ -0,0 +1,74 @@
+import org.osgi.service.component.annotations.Reference;
+
+@Component(service = { Servlet.class })
+@SlingServletResourceTypes(
+        resourceTypes = { "cards/SubjectsHomepage" },
+        methods = { "GET" })
```
`selectors` should be specified in here just like how it is in https://github.com/data-team-uhn/cards/blob/34e97eb2e16de9b5d9e571f50995e855a818d33e/heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/export/ExportEndpoint.java#L40 otherwise visiting `/Subjects` will return a HTTP 500 error. I'd recommend using a selector such `csvExport` so that visiting `/Subjects.csvExport` would manually trigger the CSV export.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportEndpoint.java:
```diff
@@ -0,0 +1,74 @@
+            return;
+        }
+
+        final ExportConfigDefinition config = this.configs.stream().map(ExportConfig::getConfig).findFirst()
+                .orElse(null);
```
It would be nice to be able to call different CSV export configurations instead of only being able to call the first one that is found. For example, we are able to create many _PROMs import_ configurations, assigning each one a unique `name`. Then, upon manual export, we can specify which _PROMs import_ configuration that we wish to run. For example, if we had a _PROMs import_ configuration with a `name` of `importConf001` we can manually trigger it by hitting `http://localhost:8080/Subjects.importTorch?config=importConf001`. In the same manner, if we create a _Survey csv export_ task with a `name` of `csvExportConf001`, we should be able to manually trigger it by hitting `http://localhost:8080/Subjects.csvExport?conf=csvExportConf001`.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,44 @@
+    /** The local path where you want to save file with exported surveys. */
+    String SAVE_PATH = "";
+
+    @AttributeDefinition(name = "Frequency in days", description = "Count of days how often should the task be invoked")
```
```suggestion
    @AttributeDefinition(name = "Frequency in days", description = "Time interval (days) between task invocations")
```

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,44 @@
+    @AttributeDefinition(name = "Frequency in days", description = "Count of days how often should the task be invoked")
+    int frequency_in_days() default FREQUENCY_IN_DAYS;
+
+    @AttributeDefinition(name = "Questionnaires to be export",
```
```suggestion
    @AttributeDefinition(name = "Questionnaires to be exported",
```

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,44 @@
+    String[] questionnaires_to_be_exported();
+
+    @AttributeDefinition(name = "Save path",
+            description = "The local path where it is specified to save file with exported surveys")
```
```suggestion
            description = "The local path to the directory where exported survey CSV files are to be saved")
```

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add name property for each new configuration

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add name property for each new configuration

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add name property for each new configuration

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add name property for each new configuration

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ScheduledExport.java:
```diff
@@ -0,0 +1,87 @@
+public class ScheduledExport
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScheduledExport.class);
+    private static final String SCHEDULER_JOB_PREFIX = "ScheduledExport-";
```
`ScheduledExport-` is too generic as we also do JSON exports and might need to support additional export formats in the future. How about instead using `ScheduledExport-CSV-`?

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ScheduledExport.java:
```diff
@@ -0,0 +1,87 @@
+
+    public void configRemoved(final ExportConfig removedConfig)
+    {
+        LOGGER.debug("Removed torch importer config {}", removedConfig.getConfig().name());
```
This is not part of the _Torch importer_. Perhaps this should be changed to `Removed CSV exporter config {}`?

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ScheduledExport.java:
```diff
@@ -0,0 +1,87 @@
+        options.name(SCHEDULER_JOB_PREFIX + newConfig.getConfig().name());
+        options.canRunConcurrently(true);
+
+        final Runnable importJob;
+        importJob =
+                new ExportTask(this.resolverFactory, newConfig.getConfig().frequency_in_days(),
+                        newConfig.getConfig().questionnaires_to_be_exported(),
+                        newConfig.getConfig().save_path());
+        try {
+            this.scheduler.schedule(importJob, options);
```
`importJob` is misleading as the task _exports_ data rather than _imports_ it. I suggest renaming it to either `exportJob` or `csvExportJob`?

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,123 @@
+    private String generateRequestedIdsQueryEnding()
+    {
+        String requestIdsQueryEnding = "";
+        if (this.questionnairesToBeExported != null && !this.questionnairesToBeExported.isEmpty()) {
```
If `this.questionnairesToBeExported` is empty, we will end up returning an empty string here which will imply that all questionnaires should be exported (as we don't have any `WHERE q.[jcr:uuid] =` constraints). This could be unexpected behavior as one would assume that an empty list of questionnaires would mean that no questionnaires would be exported instead of all of them.

I suggest that, instead of returning an empty string in the case of no questionnaires to be exported, return `null` and if there are questionnaires to be exported, return the full query string (ie. `SELECT * FROM [cards:Questionnaire] AS q WHERE q.[jcr:uuid] = ....`) This method could then instead be renamed to something like `generateQuestionnairesQuery`.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,123 @@
+    private Set<Resource> getQuestionnaires(String requestedIdsQuery)
+    {
+        try (ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(null)) {
+            String query = "SELECT * FROM [cards:Questionnaire] AS q" + requestedIdsQuery;
```
Get the entire query from the `generateQuestionnairesQuery` method (please see my other comment) and if `generateQuestionnairesQuery` returns `null`, return a `Collections.emptySet()`.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,123 @@
+    public void run()
+    {
+        Set<Resource> questionnaires = this.getQuestionnaires(generateRequestedIdsQueryEnding());
+        final Date lastExportDate = new Date(System.currentTimeMillis() - (this.frequencyInDays * DAY_IN_MS));
```
I'm not sure if this is what we want. If we do this, we risk missing data due to jitter in the scheduler. For example, if the export task runs today at 12:00:00 PM it would export all data created after 12:00:00 PM yesterday. If a Form is created today at 12:00:01 PM and tomorrow the export tasks runs at 12:00:02 PM (2 second delay due to schedule jitter), said Form would not be exported as the scheduler would only export the Forms that were created on the previous day after 12:00:02 PM thus missing the Form created at 12:00:01 PM.

Perhaps a good solution would be to schedule this job to run slightly after midnight (maybe 00:15:00) and CSV-export the data from the previous `this.frequencyInDays` days? For example, if this task runs at _2022-09-15 00:15:00_ and `this.frequencyInDays` is set to `1`, all data created at any time on _2022-09-14_ would be CSV-exported. Similarly, if this task runs at _2022-09-15 00:15:00_ and `this.frequencyInDays` is set to `2`, all data created at any time on _2022-09-13_ or _2022-09-14_ would be CSV exported.

@marta- ?

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

changed the logic of filtering the searched forms by date and questionnaire id

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

changed the logic of filtering the searched forms by date and questionnaire id

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

changed the logic of filtering the searched forms by date and questionnaire id

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

changed the logic of filtering the searched forms by date and questionnaire id

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    public void run()
+    {
+        Set<Resource> questionnaires = this.getQuestionnaires(generateQuestionnairesQuery());
```
Given that `generateQuestionnairesQuery` and `getQuestionnaires` are _private_ methods and that `generateQuestionnairesQuery` is only ever called here and `getQuestionnaires` is only ever called with `generateQuestionnairesQuery()` as an argument, it would be better to move the call to `generateQuestionnairesQuery()` to inside the body of the `getQuestionnaires` method.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ScheduledExport.java:
```diff
@@ -0,0 +1,87 @@
+            return;
+        }
+
+        final String schedule = "0 0 */" + 24 * newConfig.getConfig().frequency_in_days() + " ? * *";
```
This isn't a valid Quartz Cron string. For example, if `newConfig().getConfig().frequency_in_days()` resolves to `1` the `schedule` string is set to `0 0 */24 ? * *`. This shows as invalid on https://www.freeformatter.com/cron-expression-generator-quartz.html. Furthermore, we should be able to set a time other than midnight when the CSV export task runs. For example, one might want to specify something like _"export every 3 days at 00:30:00"_ or _"export every 1 day at 02:00:00"_.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,135 @@
+        return Collections.emptySet();
+    }
+
+    public Date getLastExportDate()
```
Does this need to be _public_? Also, I'd recommend renaming this method to `getExportStartDate`.

sofi2002sofi committed:
CARDS-1881:Periodical export task that serializes new forms as CSVs

add the functionality of specifying the export time

sofi2002sofi committed:
CARDS-1881:Periodical export task that serializes new forms as CSVs

add the functionality of specifying the export time

sofi2002sofi committed:
CARDS-1881:Periodical export task that serializes new forms as CSVs

add the functionality of specifying the export time

sofi2002sofi committed:
CARDS-1881:Periodical export task that serializes new forms as CSVs

add the functionality of specifying the export time

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,136 @@
+        String query = null;
+        if (this.questionnairesToBeExported != null && !this.questionnairesToBeExported.isEmpty()) {
+            ListIterator<String> export = this.questionnairesToBeExported.listIterator();
+            StringBuilder conditionQuery = new StringBuilder("SELECT * FROM [cards:Questionnaire] AS q "
+                    + "WHERE q.[jcr:uuid] = '" + export.next() + "'");
+            while (export.hasNext()) {
+                conditionQuery.append(" or q.[jcr:uuid] = '").append(export.next()).append("'");
```
The problem with this is that the administrator user would almost certainly configure the _Questionnaires_ to be exported by their names (eg. `AUDITC`, `GAD7`, etc...) as opposed to their `jcr:uuid` values (which are randomly generated during the JCR initialization) therefore the generated queries would contain `WHERE q.[jcr:uuid] = 'Questionnaire Name'` (eg. `WHERE q.[jcr:uuid] = 'AUDITC'`) and therefore would match nothing.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,136 @@
+        final String modifiedBeforeDate = simpleDateFormat.format(new Date());
+        final File csvFile = new File(this.savePath + "/ExportedForms_" + modifiedAfterDate + ".csv");
+        try (FileWriter writer = new FileWriter(csvFile)) {
+            for (Resource questionnaire : questionnaires) {
+                final String csvPath = String.format(
+                        questionnaire.getPath() + ".data.dataFilter:modifiedAfter=%s.dataFilter:modifiedBefore=%s.csv",
```
Given that, as I have noted in https://github.com/data-team-uhn/cards/pull/1161#discussion_r981701835, the set of _Questionnaires_ to be exported would be specified in terms of their names (eg. `AUDITC`, `GAD7`, etc...) and not their unpredictable `jcr:uuid` values and that all `cards:Questionnaire` nodes are located in the JCR under `/Questionnaires`, we can remove the `getQuestionnaires` and `generateQuestionnairesQuery` methods and simply iterate over `this.questionnairesToBeExported` and replace `questionnaire.getPath()` with `"/Questionnaires/" + (the .next() item from this.questionnairesToBeExported)`.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,136 @@
+        final SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
+        final String modifiedAfterDate = simpleDateFormat.format(getExportStartDate());
+        final String modifiedBeforeDate = simpleDateFormat.format(new Date());
+        final File csvFile = new File(this.savePath + "/ExportedForms_" + modifiedAfterDate + ".csv");
+        try (FileWriter writer = new FileWriter(csvFile)) {
```
The problem with only opening one CSV file and then writing all CSV exports of different _Questionnaire_ types to it is that the column headers for all but the first exported _Questionnaire_ will not be at the top of the CSV file. For example, our CSV file might look like

| Questionnaire 1 - Question 1 | Questionnaire 1 - Question2 | Questionnaire 1 - Question 3 |
| -------------------------------------- | -------------------------------------- | -------------------------------------- |
| some answer                         | another answer                      | another answer                      |
| some answer                         | another answer                      | another answer                      |
| some answer                         | another answer                      | another answer                      |
| Questionnaire 2 - Question 1 | Questionnaire 2 - Question2 | Questionnaire 2 - Question 3 |
| some answer                         | another answer                      | another answer                      |
| some answer                         | another answer                      | another answer                      |
| some answer                         | another answer                      | another answer                      |

while instead we would want _Questionnaire 1_ and _Questionnaire 2_ to be exported to different CSV files. Therefore, we should use a different `csvFile` and `writer` for each iterated _Questionnaire_ and include the _Questionnaire_ name in the exported file path (eg. `this.savePath + "/ExportedForms_" + questionnaireName + "_" + modifiedAfterDate + "_" + modifiedBeforeDate + ".csv"`).

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change search parameter from questionnaire id to name; implement dividing forms of different questionnaire types into separate files

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change search parameter from questionnaire id to name; implement dividing forms of different questionnaire types into separate files

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change search parameter from questionnaire id to name; implement dividing forms of different questionnaire types into separate files

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change search parameter from questionnaire id to name; implement dividing forms of different questionnaire types into separate files

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,53 @@
+        description = "Configuration for the Surveys csv exporter")
+public @interface ExportConfigDefinition
+{
+    /** How often should the task be invoked. */
+    int FREQUENCY_IN_DAYS = 1;
```
```suggestion
    /** Default value of how often the task should be invoked. */
    final int FREQUENCY_IN_DAYS = 1;
```

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,53 @@
+    /** How often should the task be invoked. */
+    int FREQUENCY_IN_DAYS = 1;
+
+    /** At which time should the task be invoked. */
+
+    String EXPORTING_TIME = "00:00:00";
```
```suggestion
    /** Default value of the time which the task should be invoked. */
    String EXPORTING_TIME = "00:00:00";
```

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,53 @@
+
+    String EXPORTING_TIME = "00:00:00";
+
+    /** The local path where you want to save file with exported surveys. */
+    String SAVE_PATH = ".";
```
```suggestion
    /** Default value of the local path where exported survey CSV files should be saved to. */
    String SAVE_PATH = ".";
```

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,53 @@
+    @AttributeDefinition(name = "Name", description = "Configuration name")
+    String name();
+
+    @AttributeDefinition(name = "Frequency in days", description = "Time interval (days) between task invocations")
+    int frequency_in_days() default FREQUENCY_IN_DAYS;
+    @AttributeDefinition(name = "Exporting time", description = "Time of task invocations (in format hh:mm:ss)")
+    String exporting_time() default EXPORTING_TIME;
```
```suggestion
    @AttributeDefinition(name = "Frequency in days", description = "Time interval (days) between task invocations")
    int frequency_in_days() default FREQUENCY_IN_DAYS;

    @AttributeDefinition(name = "Exporting time", description = "Time of task invocations (in format hh:mm:ss)")
    String exporting_time() default EXPORTING_TIME;
```

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportEndpoint.java:
```diff
@@ -0,0 +1,83 @@
+        final Runnable exportJob = new ExportTask(this.resolverFactory, config.frequency_in_days(),
+                config.questionnaires_to_be_exported(), config.save_path());
+        final Thread thread = new Thread(exportJob);
+        thread.start();
```
If we are able to start a thread of `exportJob`, we should notify the caller by returning a _HTTP 200 OK_ response and some text stating that the export is now in progress.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,92 @@
+        final String modifiedBeforeDate = simpleDateFormat.format(new Date());
+        try (ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(null)) {
+            for (String questionnaire : this.questionnairesToBeExported) {
+                final File csvFile = new File(this.savePath + "/ExportedForms_" + questionnaire + "_"
+                        + modifiedAfterDate + "_" + modifiedBeforeDate + ".csv");
```
I think it would be nicer if the dates in the CSV file name were _inclusive_. For example, the file `ExportedForms_GAD7_2022-10-05_2022-10-07.csv` would only contain GAD7 Forms that were created/modified on _2022-10-06_ but I think that many people looking at the filename would assume it to contain data from _2022-10-05_, _2022-10-06_, and _2022-10-07_.

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change naming way of exported files

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change naming way of exported files

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change naming way of exported files

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

change naming way of exported files

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,100 @@
+    public void run()
+    {
+        final SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
+        final String modifiedAfterDate = simpleDateFormat.format(getPastDate(this.frequencyInDays + 1));
```
Upon further testing and examination of https://github.com/data-team-uhn/cards/blob/5bd1edb22a32dad7b264eea6f7d009d72fee9c3b/modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/DataProcessor.java#L249-L254 we'll need to set `modifiedAfterDate` to the first date that we want to be included in the CSV export as `DataProcessor.java` uses `[jcr:lastModified] >=` (as opposed to `[jcr:lastModified] >`) for handling the `modifiedAfter` filter. For example, if we want to export all data created on and after _2022-10-09_, we would use `2022-10-09` for the value of `modifiedAfterDate` as opposed to `2022-10-08`.

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

fix calculating of the start exporting date

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

fix calculating of the start exporting date

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

fix calculating of the start exporting date

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

fix calculating of the start exporting date

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ScheduledExport.java:
```diff
@@ -0,0 +1,93 @@
+        final int hour = Integer.parseInt(time[0].trim());
+        final int min = Integer.parseInt(time[1].trim());
+        final int second = Integer.parseInt(time[2].trim());
+        final String schedule = second + SPACE + min + SPACE + hour + " */" + newConfig.getConfig().frequency_in_days()
+                + " * ?";
```
Due to the `*/FREQUENCY_IN_DAYS` term, this task will run every `newConfig.getConfig().frequency_in_days()` days _starting on the first of the month_. This is a problem as, for example, if `newConfig.getConfig().frequency_in_days()` returns `4`, we can check with https://www.freeformatter.com/cron-expression-generator-quartz.html and see that the task will execute on October 29th 2022 and then again on November 1st 2022 - which is 3 days away from October 29th 2022, not 4 days.

Perhaps it would be better to simply have an OSGi parameter for specifying the Quartz cron schedule string and passing that string directly through to `this.scheduler.EXPR`? @sdumitriu

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add new schedule OSGi parameter

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add new schedule OSGi parameter

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add new schedule OSGi parameter

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

add new schedule OSGi parameter

IntegralProgrammer commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -70,7 +70,9 @@
         "io.uhndata.cards.proms-backend:TorchImporter=[proms-import-backend]",
-        "io.uhndata.cards.proms-backend:MetricLogger=[cards-metrics]"
+        "io.uhndata.cards.proms-backend:MetricLogger=[cards-metrics]",
+        "io.uhndata.cards.proms-backend:SlackNotifications=[cards-metrics]",
```
```suggestion
```

The `SlackNotifications` _service user_ does not need to be configured here as that _service user_ is only used by the periodic Slack notifications task that reads the values of the `/Metrics/*` JCR nodes and generates a performance update message for Slack.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/export/ExportTask.java:
```diff
@@ -0,0 +1,99 @@
+            final String endModificationDate = simpleDateFormat.format(getPastDate(1));
+            timePeriod = modifiedAfterDate + "_" + endModificationDate;
+        }
+        try (ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(null)) {
```
It would make for more organized code to use a sub-service.
```suggestion
       Map<String, Object> params = new HashMap<>();
       params.put(ResourceResolverFactory.SUBSERVICE, "CsvExport");
        try (ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(params)) {
```
Of course you'll need to import `java.util.Map` and `java.util.HashMap` at the beginning of the file.

IntegralProgrammer commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -70,7 +70,9 @@
-        "io.uhndata.cards.proms-backend:MetricLogger=[cards-metrics]"
+        "io.uhndata.cards.proms-backend:MetricLogger=[cards-metrics]",
+        "io.uhndata.cards.proms-backend:SlackNotifications=[cards-metrics]",
+        "io.uhndata.cards.proms-backend=[sling-readall]"
```
After following my other comment about using a sub-service, the following change would be required:
```suggestion
        "io.uhndata.cards.proms-backend:CsvExport=[sling-readall]"
```
Additionally, instead of using `sling-readall` you might want to create a `csv-export` _service user_ that is restricted to only reading the JCR nodes that it needs to read.

IntegralProgrammer reviewed (changes requested):
I did some quick testing and everything seems to be working well. I noted just a few minor improvements that can be done. Lastly, this PR introduces the files `AIP.xml`, `ED.xml`, `OAIP.xml`, and `OED.xml`. I haven't been using those Questionnaires, however, for testing the PR. Do these files need to be in this PR or would it be better for them to be introduced in another PR?

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

create separate service user with appropriate rights for exporting forms in csv format

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

create separate service user with appropriate rights for exporting forms in csv format

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

create separate service user with appropriate rights for exporting forms in csv format

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

create separate service user with appropriate rights for exporting forms in csv format

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

delete unused files, change service user for the CsvExport sub-service

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

delete unused files, change service user for the CsvExport sub-service

IntegralProgrammer commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -62,15 +62,18 @@
-        "create group TrustedUsers \n\n set ACL for TrustedUsers \n   deny rep:write on /Subjects \n   allow jcr:removeChildNodes on /Forms \n     deny jcr:removeNode on /Forms restriction(rep:ntNames,cards:Form,cards:Subject) \n     allow jcr:removeNode on /Forms restriction(cards:createdBy) \n    deny jcr:all on /Forms restriction(cards:unsubmittedForms) \n end "
+        "create group TrustedUsers \n\n set ACL for TrustedUsers \n   deny rep:write on /Subjects \n   allow jcr:removeChildNodes on /Forms \n     deny jcr:removeNode on /Forms restriction(rep:ntNames,cards:Form,cards:Subject) \n     allow jcr:removeNode on /Forms restriction(cards:createdBy) \n    deny jcr:all on /Forms restriction(cards:unsubmittedForms) \n end ",
+        // Exporting forms in csv format
+        "create service user csv-export \n set ACL on /Questionnaires,/Forms \n   allow jcr:read for csv-export \n end"
```
The `csv-export` _service user_ will also need `jcr:read` access to `/Subjects` and `/SubjectTypes`.

```suggestion
        "create service user csv-export \n set ACL on /Questionnaires,/Forms,/Subjects,/SubjectTypes \n   allow jcr:read for csv-export \n end"
```

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

extend permisions of csv-export service user

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

extend permisions of csv-export service user

IntegralProgrammer reviewed (approved):
Looks good! :heavy_check_mark: Here's how I tested it:

1. Build this `CARDS-1881` branch with `mvn clean install`.
2. Start CARDS with `./start_cards.sh --project cards4proms --dev`.
3. Login to http://localhost:8080 as `admin`:`admin` and visit http://localhost:8080/system/console/configMgr and create a new _Survey csv export_ configuration with the following parameters:
  - `Name`: `csvTest`
  - `Frequency in days`: `1`
  - `Export schedule`: `0 0 0 * * ? *`
  - `Questionnaires to be exported`: `Patient information`
  - `Save path`: `/file/system/path/to/csv/export/directory`
4. Visit http://localhost:8080/Subjects.csvExport?config=csvTest.
5. An `ExportedForms_Patient information` CSV file should be created that contains only the header row.
6. Create a new _Patient information_ Form and populate it with data.
7. Visit http://localhost:8080/Subjects.csvExport?config=csvTest.
8. The `ExportedForms_Patient information` CSV file should still only contain the header row.
9. With the _Patient information_ Form still open for editing, open a new tab and visit http://localhost:8080/bin/browser.html and set the `jcr:lastModified` property of that Form node to yesterday's date.
10. Visit http://localhost:8080/Subjects.csvExport?config=csvTest.
11. The `Exported_Forms_Patient information` CSV file should now contain a data row for the exported _Patient information_ form data.
12. Repeat this process for multiple _Patient information_ Forms. Ensure that each Form with a `jcr:lastModified` value of _yesterday_ is being written to a row in the CSV export of today's data.
13. Add other Questionnaires to the `Questionnaire to be exported` configuration parameter. Ensure that they are being exported to different files as well.
14. Set the `Export schedule` parameter to `0 * * * * ? *`. Verify that the export task is running once per minute.
15. Modify the `Frequency in days` parameter and verify that the expected Forms (given their `jcr:lastModified` property value) are being exported to CSV files.
16. Create another _Survey csv export_ task (call it `csvTest2`) and ensure that it, along with the previous _Survey csv export_ task (which we called `csvTest`) are both working as expected.
17. Delete the first `csvTest` export task. Ensure that the periodic CSV data export stops but that the `csvTest2` export task still continues.
18. Finally, delete the `csvTest2` export task and ensure that the data CSV data export stops.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
PROBLEM: works well for PROMS, not PREMS
In PREMS http://localhost:8080/system/console/configMgr is missing `Surveys csv export`

![image](https://user-images.githubusercontent.com/23243244/201136889-0d555a94-77c0-4d2e-80cc-17f8e2e6ae25.png)

[PROMS.pdf](https://github.com/data-team-uhn/cards/files/9982295/PROMS.pdf)
[PREMS.pdf](https://github.com/data-team-uhn/cards/files/9982298/PREMS.pdf)

marta- commented:
@IntegralProgrammer Update re code reorganization: move everything in a core module named `scheduled-csv-export`, and pull this module in only in the `prems` project.

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Move from proms-resources to prems-resources

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Move from proms-resources to prems-resources

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Move the proms-resources feature.json configurations to the prems-resources feature.json configuration file

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Move the proms-resources feature.json configurations to the prems-resources feature.json configuration file

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Fix syntax error in prems-resources feature.json file

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Fix syntax error in prems-resources feature.json file

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Move the CSV exporter into its own core module under modules/scheduled-csv-export

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Move the CSV exporter into its own core module under modules/scheduled-csv-export

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Modify start_cards.sh to automatically load the scheduled-csv-export module whenever the cards4prems project is specified

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Modify start_cards.sh to automatically load the scheduled-csv-export module whenever the cards4prems project is specified

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Allow for the scheduled CSV exporter to also be enabled through the --csv_export flag for start_cards.sh

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Allow for the scheduled CSV exporter to also be enabled through the --csv_export flag for start_cards.sh

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Update distribution/docker_entry.sh so that the scheduled-csv-export module is enabled whenever --cards_project cards4prems is specified

IntegralProgrammer committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

Update distribution/docker_entry.sh so that the scheduled-csv-export module is enabled whenever --cards_project cards4prems is specified

marta- commented:
This needs to be retested post refactoring. Nothing should have changed functionality-wise.

IntegralProgrammer commented:
> Looks good! heavy_check_mark Here's how I tested it:
> 
>     1. Build this `CARDS-1881` branch with `mvn clean install`.
> 
>     2. Start CARDS with `./start_cards.sh --project cards4proms --dev`.
> 
>     3. Login to http://localhost:8080 as `admin`:`admin` and visit http://localhost:8080/system/console/configMgr and create a new _Survey csv export_ configuration with the following parameters:
> 
> 
>     * `Name`: `csvTest`
> 
>     * `Frequency in days`: `1`
> 
>     * `Export schedule`: `0 0 0 * * ? *`
> 
>     * `Questionnaires to be exported`: `Patient information`
> 
>     * `Save path`: `/file/system/path/to/csv/export/directory`
> 
> 
>     4. Visit http://localhost:8080/Subjects.csvExport?config=csvTest.
> 
>     5. An `ExportedForms_Patient information` CSV file should be created that contains only the header row.
> 
>     6. Create a new _Patient information_ Form and populate it with data.
> 
>     7. Visit http://localhost:8080/Subjects.csvExport?config=csvTest.
> 
>     8. The `ExportedForms_Patient information` CSV file should still only contain the header row.
> 
>     9. With the _Patient information_ Form still open for editing, open a new tab and visit http://localhost:8080/bin/browser.html and set the `jcr:lastModified` property of that Form node to yesterday's date.
> 
>     10. Visit http://localhost:8080/Subjects.csvExport?config=csvTest.
> 
>     11. The `Exported_Forms_Patient information` CSV file should now contain a data row for the exported _Patient information_ form data.
> 
>     12. Repeat this process for multiple _Patient information_ Forms. Ensure that each Form with a `jcr:lastModified` value of _yesterday_ is being written to a row in the CSV export of today's data.
> 
>     13. Add other Questionnaires to the `Questionnaire to be exported` configuration parameter. Ensure that they are being exported to different files as well.
> 
>     14. Set the `Export schedule` parameter to `0 * * * * ? *`. Verify that the export task is running once per minute.
> 
>     15. Modify the `Frequency in days` parameter and verify that the expected Forms (given their `jcr:lastModified` property value) are being exported to CSV files.
> 
>     16. Create another _Survey csv export_ task (call it `csvTest2`) and ensure that it, along with the previous _Survey csv export_ task (which we called `csvTest`) are both working as expected.
> 
>     17. Delete the first `csvTest` export task. Ensure that the periodic CSV data export stops but that the `csvTest2` export task still continues.
> 
>     18. Finally, delete the `csvTest2` export task and ensure that the data CSV data export stops.

:point_up: Re-tested this PR using these instructions after applying the refactoring / rebase. All looks good. @sashaandjic Please test and confirm.

sashaandjic reviewed (approved):
tested

marta- commented:
We need the ability to export the answer *labels* instead of the *values* in the CSVs. The values are exported by default, while for exporting labels we need to include the `labels` processor in the export url, e.g. `/Questionnaires/OED.labels.<other_processors>.csv`.

To do: add a labels check-mark box to the OSGi configuration for the Scheduled CSV exporter. If that box is enabled, then `.labels` would be appended to the data export URL that is queried for obtaining the data to be written to the CSV file.

sofi2002sofi committed:
CARDS-1881: Enable labels processor in CSV exporting

sofi2002sofi committed:
CARDS-1881: Enable labels processor in CSV exporting

veronikaslc commented:
There are 10 new `.java` classes in the PR: 2 copies of `ExportConfigDefinition`, 2 copies of `ExportConfig`, etc. In short `io.uhndata.cards.proms.internal.export` package contains a full copy of `io.uhndata.cards.scheduledcsvexport`. Not sure we need 2 copies, we don't use them both. If yes - we definetly can avoid code duplication. If they were forgotten to be deleted at the step of moving to it's own `scheduled-csv-export` module, all changes to `proms-resources/feature/src/main/features/feature.json` have to be reverted as well.

veronikaslc commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ScheduledExport.java:
```diff
@@ -0,0 +1,86 @@
+            // If the scheduler is not bound, it's OK to do nothing now, the method will be called during activation.
+            return;
+        }
+        final String schedule = newConfig.getConfig().export_schedule();
```
```suggestion

        final ExportConfigDefinition configDef = newConfig.getConfig();
        LOGGER.debug("Added csv export configuration {}", configDef.name());
        final String schedule = configDef.export_schedule();
```

veronikaslc replied:
We may use more logging as well as avoid `newConfig.getConfig()` repeated 6 times. I know it's the way it was done in `NightlyImport` where it's repeated 12 times, but we still can optimise at least locally.

veronikaslc commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ExportConfig.java:
```diff
@@ -0,0 +1,42 @@
+    private ExportConfigDefinition config;
+
+    @Activate
+    protected void activate(final ExportConfigDefinition config, ComponentContext componentContext)
```
```suggestion
    protected void activate(final ExportConfigDefinition config, final ComponentContext componentContext)
```

veronikaslc commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ExportEndpoint.java:
```diff
@@ -0,0 +1,85 @@
+
+        //Ensure that this can only be run when logged in as admin
+        final String remoteUser = request.getRemoteUser();
+        if (remoteUser == null || !remoteUser.equals(ADMIN)) {
```
Our login system authorises `Admin` cred as a `remoteUser` and here it kicks out with `403`. We have to either disallow uppersase admin username or - better practice - we have to do case-insensitive comparison.

veronikaslc replied:
```suggestion
        if (remoteUser == null || !remoteUser.toLowerCase(Locale.ROOT).equals(ADMIN)) {
```

veronikaslc reviewed (commented):
Tested. Labels are exported as expected.

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

clean up the proms-resources folder, code optimization

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

clean up the proms-resources folder, code optimization

veronikaslc commented:
We still have changes to `proms-resources/feature/src/main/features/feature.json`

marta- commented:
> We still have changes to `proms-resources/feature/src/main/features/feature.json`

Clarification: this change needs to be moved to the `prems` feature file so that the csv export task will be enabled in `prems`.

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

move proms features to prems

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

move proms features to prems

veronikaslc commented:
@marta- I might be wrong but I have doubts we need/use changes introdused in `prems-resources/feature/src/main/features/feature.json` at all.

marta- commented:
> @marta- I might be wrong but I have doubts we need/use changes introdused in `prems-resources/feature/src/main/features/feature.json` at all. Csv export task is already enabled in `prems` [here](https://github.com/data-team-uhn/cards/pull/1161/files#diff-aa150e59f8667e1f84c94053c306bf9f2cbd2884551410e6ee7ad490c3195615R56) and we have its own `feature.json` in `scheduled-csv-export` module.

I see, it's pulled in for the cards4prems project in `start_cards.sh` and `docker_entry.sh`. Indeed we can just remove those changes.

marta- commented:
Looking at the commit history, I see 4 commits from August 4 for issue 1781 that are probably another imerge/rebase artifact and need to be removed via `git rebase --interactive` @sofi2002sofi .

veronikaslc commented:
@marta- I don't think commits from 1781 are artifacts. The 1881 issue specs state:
>The the first version of the implementation of this export task could branch off of the CARDS-1781 branch for easier testing.

veronikaslc reviewed (changes requested):
changes to `prems-resources/feature/src/main/features/feature.json` has to be reverted.

sofi2002sofi committed:
Revert "CARDS-1881: Periodical export task that serializes new forms as CSVs"

This reverts commit 7a8dff32e479b870029107b4349d7d6a5e479a7f.

sofi2002sofi committed:
Revert "CARDS-1881: Periodical export task that serializes new forms as CSVs"

This reverts commit 7a8dff32e479b870029107b4349d7d6a5e479a7f.

veronikaslc commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ExportEndpoint.java:
```diff
@@ -0,0 +1,86 @@
+                .filter(c -> configName.equals(c.getConfig().name()))
+                .map(ExportConfig::getConfig)
+                .findFirst().orElse(null);
+        final Runnable exportJob = new ExportTask(this.resolverFactory, config.frequency_in_days(),
```
@sdumitriu I see from the line above that `config` is nullable. Could `config.frequency_in_days()` and such potentially throw an `NPE`?

veronikaslc replied:
(Above question also affects for `ScheduledExport.java`)

sdumitriu replied:
Yes, but I see this is already addressed.

veronikaslc commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ScheduledExport.java:
```diff
@@ -0,0 +1,88 @@
+
+        final Runnable csvExportJob;
+        csvExportJob =
+                new ExportTask(this.resolverFactory, newConfig.getConfig().frequency_in_days(),
```
`newConfig.getConfig()` is already defined on the line 58 as `configDef`, should be used here.

sofi2002sofi committed:
Merge branch 'CARDS-1881' of https://github.com/data-team-uhn/cards into CARDS-1881

veronikaslc commented:
Left 2 comments (now can be seen before rebase commits)

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

use a declared variable instead of an inline one

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

NPE fix

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/scheduled-csv-export/pom.xml:
```diff
@@ -0,0 +1,96 @@
+    <dependency>
+      <groupId>javax.jcr</groupId>
+      <artifactId>jcr</artifactId>
+    </dependency>
```
Not used, remove.

```suggestion
```

sdumitriu commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,60 @@
+    String export_schedule() default NIGHTLY_EXPORT_SCHEDULE;
+
+    @AttributeDefinition(name = "Questionnaires to be exported",
+            description = "List of questionnaires specified to be exported periodically")
```
```suggestion
            description = "List of questionnaires specified to be exported periodically, as full paths")
```

sdumitriu commented on modules/scheduled-csv-export/src/main/java/io/uhndata/cards/scheduledcsvexport/ExportConfigDefinition.java:
```diff
@@ -0,0 +1,60 @@
+    String[] questionnaires_to_be_exported();
+
+    @AttributeDefinition(name = "Save path",
+            description = "The local path to the directory where exported survey CSV files are to be saved")
```
```suggestion
            description = "The local disk path to the directory where exported survey CSV files are to be saved")
```

sofi2002sofi committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs

remove useless dependency, edit attribute definition description

sdumitriu committed:
[cleanup] Applied codestyle

sdumitriu committed:
CARDS-1881: Periodical export task that serializes new forms as CSVs
Change the questionnaires configuration from just node names to actual paths
